### PR TITLE
Added a buildrequestmerger that merges things before they hit the database

### PR DESCRIPTION
**Pull request by campos-ddc**
This will also require some changes to Katana configs, since now we are using a list of `merge_properties` to check, instead of python merge functions.

The plan is to test this in katana staging and keep an eye on logs to see if this "merge-on-add" covers all of our current cases.

### CODE REVIEW

File: master/buildbot/config.py [hunk 1]
```diff
@@ -818,6 +819,10 @@ def __init__(self, name=None, slavename=None, startSlavenames=None, slavenames=N
         self.tags = tags
         self.build_tags = build_tags
 
+        # Additional properties that must match to allow buildrequests of
+        # this config to be merged
+        self.mergeProperties = mergeProperties or []
```
Comment (mariangemarcano) [inline master/buildbot/config.py:824]: I think this introduces a new way of checking for merges that might not be compatible with the merging functions (for example this could break partition build request for parallel tests where a builder may or not have the partition id, one example is  "Test IntegrationTests - Windows64Editor" builder

I think best case we could try to reuse the logic here: FindPreviousSuccessBuildMixin

File: master/buildbot/process/buildrequestmerger.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        start = time.time()
+
+        buildsetLog = {
+            'name': 'addBuildset',
+            'description': 'Log merges done while adding new buildsets',
+            'sourcestampsetid': sourcestampsetid,
+            'builderNames': builderNames,
+        }
+
+        codebase, branch, revision = self._getSourceStampInfo(sourcestampsetid)
```
Comment (mariangemarcano) [inline master/buildbot/process/buildrequestmerger.py]: not supporting multiple codebases

File: master/buildbot/process/buildrequestmerger.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+        # Check if relevant properties match
+        for otherBuildsetid, otherBrid in matchingBrids:
+            otherProperties = self.properties_cache.get(otherBuildsetid)
+            if self._propertiesMatch(properties, otherProperties,
+                                     builder.config.mergeProperties):
+                return otherBrid
+        return None
+
+    def _getBuildRequestIdsMatchingSourceStamp(self, builderName, priority,
+                                               codebase, branch, revision):
```
Comment (mariangemarcano) [inline master/buildbot/process/buildrequestmerger.py]: I think we should use this instead it has the multiple codebase support and add support to get only the running builds
```
            prevBuildRequests = yield master.db.buildrequests \
                .getBuildRequestsBySourcestamps(buildername=build.builder.config.name,
                                                sourcestamps = build_sourcestamps)
```

File: master/buildbot/process/buildrequestmerger.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+            # Join on same sourcestamp (codebase, branch, revision)
+            #     .join(bs_tbl, bs_tbl.c.id == breq_tbl.c.buildsetid) \
+            #     .join(ss_tbl, ss_tbl.c.sourcestampsetid == bs_tbl.c.sourcestampsetid) \
+            #     .where(
+            #         ss_tbl.c.codebase == codebase,
+            #         ss_tbl.c.branch == branch,
+            #         ss_tbl.c.revision == revision,
+            #     )
+            q = sa.select(breq_tbl.c.buildsetid, breq_tbl.c.id) \
+                .where(breq_tbl.c.complete == 0) \
```
Comment (mariangemarcano) [inline master/buildbot/process/buildrequestmerger.py]: if we filter by complete builds, then we won't be merging in progress builds? that's what it is in the logic of the merging function that was disabled

**Review by mariangemarcano — CHANGES_REQUESTED** [review]
Some comments about the solution: 
- New way of checking merges that might not be compatible with the functions configured (it could break parallel test support).
- No support for multiple codebases.
- New builds (without revision id would need to be queue, and that means the merging would be skip)

File: master/buildbot/config.py (hunk 1 above)
Comment (campos-ddc) [reply master/buildbot/config.py:824]: All existing merging functions use the MatchProperty function, which basically compares properties and ensure they are the same. It would be easy to reproduce that using this list of property names and will not need any python function calls.

File: master/buildbot/process/buildrequestmerger.py (hunk 2 above)
Comment (campos-ddc) [reply master/buildbot/process/buildrequestmerger.py]: What happens when there are multiple codebases?

Will we have several entries in the `sourcestamps` table with the same `sourcestampsetid`?

File: master/buildbot/process/buildrequestmerger.py (hunk 3 above)
Comment (campos-ddc) [reply master/buildbot/process/buildrequestmerger.py]: I want to avoid using existing queries to make sure we are using simple ones and not affecting the rest of the code.

For example, this method you linked does a lot of queries involving the `sourcestampsets` table, which is a useless table that only contains ids.

But it will serve as a starting point for writing a multiple-codebase query myself :)

File: master/buildbot/process/buildrequestmerger.py (hunk 4 above)
Comment (campos-ddc) [reply master/buildbot/process/buildrequestmerger.py]: Yes, and that is what we are trying to bring back (using a new code that hopefully won't slow down Katana as much as the old one)

File: master/buildbot/process/buildrequestmerger.py (hunk 2 above)
Comment (mariangemarcano) [reply master/buildbot/process/buildrequestmerger.py]: yes, that's correct

File: master/buildbot/config.py (hunk 1 above)
Comment (mariangemarcano) [reply master/buildbot/config.py:824]: about parallel test and self.mergeProperties, the default function don't have the partitionid property and it is created dynamically, I’m assuming that the mergeProperty list is fixed in the config and maybe not considering this case? so non-partition builds could potentially get merged

File: master/buildbot/process/buildrequestmerger.py (hunk 4 above)
Comment (campos-ddc) [reply master/buildbot/process/buildrequestmerger.py]: To make it clear, we are querying for *incomplete* builds, so we will try to merge against both pending and running builds.

File: master/buildbot/config.py (hunk 1 above)
Comment (campos-ddc) [reply master/buildbot/config.py:824]: > I’m assuming that the mergeProperty list is fixed in the config

yes

> so non-partition builds could potentially get merged

I don't see why that would happen. If you are a partitionable config you will specifically say that we must check `partition-id` property to merge. That means that parition 1s only get merged to partition 1s. If you happen to run a partitionable config without partitioning it (do we even allow that?) then your property value will be `None`, and you will only merge against other identical builds.

File: master/buildbot/process/buildrequestmerger.py (hunk 4 above)
Comment (mariangemarcano) [reply master/buildbot/process/buildrequestmerger.py]: ok I think we could start by supporting only building builds merges, as for pending builds we would need to check what would be the impact of not adding a dependency row, I think as the code it is right now things could break as we used the triggeredbybrid to do things like the downloads

File: master/buildbot/config.py (hunk 1 above)
Comment (mariangemarcano) [reply master/buildbot/config.py:824]: I'll include @tdctaz to help check this part, I think the problem could be "Test IntegrationTests - Windows64Editor" builder (with no partition set by default) tries to merge with "Test IntegrationTests - Windows64Editor" builder" (that has the extra partition property set by trigger step https://github.com/Unity-Technologies/katana/blob/katana/master/buildbot/steps/partition_trigger.py#L43)

File: master/buildbot/process/buildrequestmerger.py (hunk 4 above)
Comment (mariangemarcano) [reply master/buildbot/process/buildrequestmerger.py]: I think we need to check that case anyways, merge building builds when trigger builds from dependencies.

File: master/buildbot/process/buildrequestmerger.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+        log.msg(json.dumps(buildsetLog))
+
+        return self.master.db.buildsets.addBuildset(
+            sourcestampsetid=sourcestampsetid,
+            reason=reason,
+            properties=properties,
+            triggeredbybrid=triggeredbybrid,
+            builderNames=builderNames,
+            breqsToMerge=breqsToMerge,
```
Comment (mariangemarcano) [inline master/buildbot/process/buildrequestmerger.py]: This one needs to be adjusted as well to get the new param self.master.db.buildsets.addBuildset

File: master/buildbot/config.py (hunk 1 above)
Comment (campos-ddc) [reply master/buildbot/config.py:824]: So the top level build will have partition `None`, and the child builds will have partition `1,2,3,etc`.

I'm assuming that the child builds will have their `partition` property set when they are added to the queue, or is that only set when they start running?

Comment (mariangemarcano) [reply master/buildbot/config.py:824]: yes the `partition` property is set when added to the queue I guess if we set in the configuration for the parallel builders to compare `partitions-id` then it could work. But I would like to check this also with @tdctaz

File: master/buildbot/process/buildrequestmerger.py (hunk 4 above)
Comment (mariangemarcano) [reply master/buildbot/process/buildrequestmerger.py]: @campos-ddc ok ignored the previous comment, you right for _incomplete_ the check is ok

**Comment by tdctaz** [comment]
If we are going to add an alternative merge logic I think we should remove the possibility to assign a merge function, so there are not two different systems for the same task where each of them only handle part of the merging. So I would remove the existing merge logic from `builder.py` see `getConfiguredMergeRequestsFn` and `getMergeRequestsFn`.

Also find previous successful builds should also get updated to use the same property logic `FindPreviousSuccessBuildMixin` in `artifact.py` handles merges for Find previous successful build and Check artifact exists.

File: master/buildbot/config.py (hunk 1 above)
Comment (tdctaz) [reply master/buildbot/config.py:824]: I think this is fine, but read my general comments below - I think the artifact merge should also be updated and possible the old merge logic should be removed so there is only one merge system active.

File: master/buildbot/process/buildrequestmerger.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+            return None
+
+        # Get an initial list of all breqs that match the given source stamp data
+        matchingBrids = self._getBuildRequestIdsMatchingSourceStamp(
+            builderName, priority, codebase, branch, revision)
+
+        # Check if relevant properties match
+        for otherBuildsetid, otherBrid in matchingBrids:
+            otherProperties = self.properties_cache.get(otherBuildsetid)
+            if self._propertiesMatch(properties, otherProperties,
```
Comment (tdctaz) [inline master/buildbot/process/buildrequestmerger.py]: You could alternative to adding a new propertyMatch instead call the merge function, it looks to me like you almost have all the information you need to create build request.

File: master/buildbot/process/buildrequestmerger.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+        if 'selected_slave' in properties:
+            return None
+
+        # Get an initial list of all breqs that match the given source stamp data
+        matchingBrids = self._getBuildRequestIdsMatchingSourceStamp(
+            builderName, priority, codebase, branch, revision)
+
+        # Check if relevant properties match
+        for otherBuildsetid, otherBrid in matchingBrids:
+            otherProperties = self.properties_cache.get(otherBuildsetid)
```
Comment (tdctaz) [inline master/buildbot/process/buildrequestmerger.py]: Would this not result in a possible N+1 select if the properties are not in the cache?

File: master/buildbot/process/buildrequestmerger.py (hunk 6 above)
Comment (campos-ddc) [reply master/buildbot/process/buildrequestmerger.py]: Almost but not really. I would mean duplicating a lot of the effort that is done later on (when we eventually add all build request information to the database)

**Comment by campos-ddc** [comment]
@tdctaz yes, the plan is to eventually remove the mergeFunction stuff, but we will keep it here while testing.

When we verify that the old merges (using functions) are not used anymore (because stuff is being merged before we get there) we will remove it.

File: master/buildbot/process/buildrequestmerger.py (hunk 7 above)
Comment (campos-ddc) [reply master/buildbot/process/buildrequestmerger.py]: Yep, one query for each `buildsetid` missing from the cache.

I'll make this smarter and use `IN <id_list>`

File: master/buildbot/process/buildrequestmerger.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    def _getBuildsetsProperties(self, buildsetids):
+        properties = {}
+
+        # Fill in properties from cache
+        for buildsetid in buildsetids:
+            if buildsetid in self.properties_cache.cache:
+                properties[buildsetid] = self.properties_cache.get(buildsetid)
+
+        # Read missing properties in a single query
+        missing_buildsetids = set(buildsetids).difference(properties.keys())
```
Comment (tdctaz) [inline master/buildbot/process/buildrequestmerger.py]: Very nice 👍

**Comment by tdctaz** [comment]
@campos-ddc I still think you should update `FindPreviousSuccessBuildMixin` in `artifact.py` to use your property function and cache.

File: master/buildbot/process/buildrequestmerger.py [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+    the top level build.
+    """
+
+    # Basic list of properties that must match for a buildrequest to be merged
+    # BuilderConfigs can define additional properties (see _propertiesMatch)
+    BASE_MERGE_PROPERTIES = ['force_rebuild', 'force_chain_rebuild']
+
+    def __init__(self, master):
+        self.master = master
+        self.startbrid_cache = LRUCache(
```
Comment (mariangemarcano) [inline master/buildbot/process/buildrequestmerger.py]: I think the async version of the cache would perform better for this case, AsyncLRUCache  

> class AsyncLRUCache(LRUCache):
>     """
>     An LRU cache with asynchronous locking to ensure that in the common case of
>     multiple concurrent requests for the same key, only one fetch is performed.
>     """

File: master/buildbot/process/buildrequestmerger.py [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+        :param list(str) mergeProperties:
+            List of properties that must match
+        :return bool:
+            True if `properties` and `otherProperties` match for a list of
+            `mergeProperties` to be checked
+        """
+        if 'selected_slave' in otherProperties:
+            return False
+
+        for propertyName in self.BASE_MERGE_PROPERTIES + mergeProperties:
```
Comment (mariangemarcano) [inline master/buildbot/process/buildrequestmerger.py:146]: it doesn't look like we handle the case where mergeProperties is None

File: master/buildbot/process/buildrequestmerger.py [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+            if buildsetid in self.properties_cache.cache:
+                properties[buildsetid] = self.properties_cache.get(buildsetid)
+
+        # Read missing properties in a single query
+        missing_buildsetids = set(buildsetids).difference(properties.keys())
+        if missing_buildsetids:
+            prop_tbl = self.master.db.model.buildset_properties
+
+            def __getBuildsetsProperties(conn):
+                q = sa.select(prop_tbl.c.buildsetid, prop_tbl.c.property_name, prop_tbl.c.property_value) \
```
Comment (mariangemarcano) [inline master/buildbot/process/buildrequestmerger.py]: Overall db functions are inside buildbot.db, I think it would be best to keep that kind organization

File: master/buildbot/process/buildrequestmerger.py [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+                    breqsToMerge[builderName] = mergeBrid
+
+            buildsetLog[builderName] = {
+                'elapsed': time.time() - builderMergeStart,
+                'mergeBrid': mergeBrid
+            }
+
+        buildsetLog['elapsed'] = time.time() - start
+
+        log.msg(json.dumps(buildsetLog))
```
Comment (mariangemarcano) [inline master/buildbot/process/buildrequestmerger.py:84]: I got this error when running locally 

> raise TypeError(repr(o) + " is not JSON serializable")
> 	exceptions.TypeError: set(['z-builder1']) is not JSON serializable

**Review by mariangemarcano — COMMENTED** [review]
Overall this approach looks good, just left out a few comments

File: master/buildbot/process/buildrequestmerger.py (hunk 9 above)
Comment (campos-ddc) [reply master/buildbot/process/buildrequestmerger.py]: The problem is that I'm updating the cache manually (by reading multiple properties at once), and doing that in the AsyncLRU seemed too finicky/unsafe.

We could:
* Rewrite some of the AsyncLRU to allow us to add multiple items in one move
* Revert back to reading one cache entry from the DB at the time (using the standard miss function flow)
* Take the performance hit and use regular LRU (what I'm doing right now)

What do you think?

File: master/buildbot/process/buildrequestmerger.py (hunk 10 above)
Comment (campos-ddc) [reply master/buildbot/process/buildrequestmerger.py:146]: It should never be `None`, it defaults to `[]`

File: master/buildbot/process/buildrequestmerger.py (hunk 9 above)
Comment (mariangemarcano) [reply master/buildbot/process/buildrequestmerger.py]: @campos-ddc I think we can make `self.startbrid_cache` use AsyncLRUCache  and I think if we make sure that the `self.properties_cache` gets called in an async way, it should be ok. Probably we would need `_getMergeBrid` used with `@defer.inlineCallbacks`

File: master/buildbot/process/buildrequestmerger.py (hunk 11 above)
Comment (campos-ddc) [reply master/buildbot/process/buildrequestmerger.py]: Will do

File: master/buildbot/process/buildrequestmerger.py (hunk 12 above)
Comment (campos-ddc) [reply master/buildbot/process/buildrequestmerger.py:84]: I guess that is coming from `builderNames`.

Tests indicate that it should be receiving a list, but maybe that is wrong :P

I'll do a `sort` in it, that will improve log readability and handle sets at the same time.

File: master/buildbot/db/buildrequests.py [hunk 13]
```diff
@@ -1302,6 +1302,59 @@ def thd(conn):
...
+            #     .where(breq_tbl.c.startbrid == startbrid) \
+
+            # With the same name
+            #     .where(breq_tbl.c.buildername == builderName) \
+
+            # That were not merged themselves (only merge against one / main target)
+            #     .where(breq_tbl.c.mergebrid == None) \
+            breq_tbl = self.db.model.buildrequests
+            q = sa.select(breq_tbl.c.buildsetid, breq_tbl.c.id) \
+                .where(breq_tbl.c.startbrid == startbrid) \
```
Comment (mariangemarcano) [inline master/buildbot/db/buildrequests.py]: @campos-ddc why did you remove incomplete build filter? I think it was fine as it was before. We need to check if the logic would work, (I think if we try to merge with a finished build like this, the build might not get completed)

**Review by mariangemarcano — APPROVED** [review]

File: master/buildbot/db/buildrequests.py (hunk 13 above)
Comment (campos-ddc) [reply master/buildbot/db/buildrequests.py]: Don't remember.

Either way, all of that is getting removed since it turns out that merging only to chain is not good enough for QV.

**Comment by campos-ddc** [comment]
Will reopen when new version is ready
